src/software/ai/passing/pass_generator.cpp

```diff
@@ -0,0 +1,179 @@
+                            receiving_position, 0.02),
+                    std::to_string(debug_shapes.size()) + "pg"));
+            }
+        }
+        std::stringstream stream;
+        stream << "BP:" << std::fixed << std::setprecision(3) << best_pass.rating;
+        debug_shapes.push_back(
+            *createDebugShape(Circle(best_pass.pass.receiverPoint(), 0.05),
+                              std::to_string(debug_shapes.size()) + "pg", stream.str()));
+        LOG(VISUALIZE) << *createDebugShapes(debug_shapes);
+    }
+
+    // Generate sample passes across the field for cost visualization
+    if (passing_config_.cost_vis_config().generate_sample_passes())
+    {
+        samplePassesForVisualization(world, passing_config_, best_pass.pass);
+    }
+
+    return best_pass;
+}
+
+std::map<RobotId, std::vector<Point>> PassGenerator::sampleReceivingPositionsPerRobot(
+    const World& world, const std::vector<RobotId>& robots_to_ignore)
+{
+    std::map<RobotId, std::vector<Point>> receiving_positions_map;
+
+    const double min_sampling_std_dev =
+        passing_config_.pass_gen_min_rand_sample_std_dev_meters();
+    const double sampling_std_dev_vel_multiplier =
+        passing_config_.pass_gen_rand_sample_std_dev_robot_vel_multiplier();
+    const double sampling_center_vel_multiplier =
+        passing_config_.pass_gen_rand_sample_center_robot_vel_multiplier();
+
+    for (const Robot& robot : world.friendlyTeam().getAllRobots())
+    {
+        // Ignore robots in the ignore list
+        if (std::find(robots_to_ignore.begin(), robots_to_ignore.end(), robot.id()) !=
+            robots_to_ignore.end())
+        {
+            continue;
+        }
+
+        // Add the robot's current position to the list of sampled passes
+        auto robot_position = robot.position();
+        receiving_positions_map.insert({robot.id(), {robot_position}});
+
+        // Add the best pass from the previous iteration to the list of sampled passes
+        if (previous_best_receiving_positions_.find(robot.id()) !=
+            previous_best_receiving_positions_.end())
+        {
+            receiving_positions_map[robot.id()].push_back(
+                previous_best_receiving_positions_[robot.id()]);
+        }
+
+        // Sample random receiving positions using a normal distribution around the
+        // robot's future position with a standard deviation that scales with the robot's
+        // velocity.
+        Point sampling_center =
+            robot_position + (robot.velocity() * sampling_center_vel_multiplier);
+        double std_dev = min_sampling_std_dev +
+                         (robot.velocity().length() * sampling_std_dev_vel_multiplier);
+        std::normal_distribution x_normal_distribution{sampling_center.x(), std_dev};
+        std::normal_distribution y_normal_distribution{sampling_center.y(), std_dev};
+
+        for (unsigned int i = 0; i < passing_config_.pass_gen_num_samples_per_robot();
```

Comment:
Suggestions that you could totally ignore: 

Since we are iterating, I feel like it would make more sense if we were to iterate through some increment of time and then find a random position and appending it to the list than to just find some random points based on one position.

Reply:
That's an interesting idea that we could explore and see how it affects the best found pass, but I think it would be out of the scope of this PR. It's also worth keeping in mind that the standard deviation of the normal distribution is fairly large (specially if the robot is moving fast), so the receiving positions should already be pretty well spread out. 

You can actually visualize the sampled positions (before they are optimized) in Thunderscope by enabling `visualize_samples_passes`:

![image](https://github.com/UBC-Thunderbots/Software/assets/28585597/4d568ca9-1909-45cd-bbf6-b38443e160ed)